### PR TITLE
added additional audio options for feature extraction

### PR DESCRIPTION
**Pull request by JackTemaki**
I added 2 new feature extraction methods, one for linear spectrograms and a modified version of log-mel-spectrograms with additional options and a different scaling. 
Other additions are:

- The peak normalization can be disabled now
- A preemphasis filter can be set
- The features can be stacked as "superframes" (https://arxiv.org/pdf/1507.06947.pdf, section 2.3, https://arxiv.org/abs/1703.10135)
- the normalization can be a single value instead of a file or numpy array representing each feature dimension

The superframes can in theory be implemented as a layer, but with the existing layers I was not able to produce this behavior, and additional layers for conversion would slightly increase the memory load.

### CODE REVIEW

File: GeneratingDataset.py [hunk 1]
```diff
@@ -917,18 +917,23 @@ class ExtractAudioFeatures:
...
     :param float window_len: in seconds
     :param float step_len: in seconds
     :param int num_feature_filters:
     :param bool|int with_delta:
-    :param numpy.ndarray|str|None norm_mean: if str, will interpret as filename
-    :param numpy.ndarray|str|None norm_std_dev: if str, will interpret as filename
+    :param numpy.ndarray|str|int|float|None norm_mean: if str, will interpret as filename
+    :param numpy.ndarray|str|int|float|None norm_std_dev: if str, will interpret as filename
     :param str features: "mfcc", "log_mel_filterbank", "log_log_mel_filterbank", "raw", "raw_ogg"
+    :param dict|None feature_options: provide additional paramters for the feature function
```
Comment (albertz) [inline GeneratingDataset.py]: I guess `dict[str]`, right?

File: GeneratingDataset.py [hunk 2]
```diff
@@ -1051,16 +1078,65 @@ def get_audio_features(self, audio, sample_rate):
...
+      else:
+        feature_data /= self.norm_std_dev[None, :]
+
+    if self.join_frames is not None:
+      pad_len = self.join_frames - (feature_data.shape[0] % self.join_frames)
+      pad_len = pad_len % self.join_frames
+      new_len = feature_data.shape[0] + pad_len
+      feature_data = numpy.pad(feature_data, pad_width=((0, pad_len), (0, 0)), mode="edge")
+      feature_data = numpy.reshape(feature_data,
+                                   newshape=(int(new_len/self.join_frames), feature_data.shape[1]*self.join_frames),
```
Comment (albertz) [inline GeneratingDataset.py]: There are some spaces missing (around any operators).

Comment (albertz) [reply GeneratingDataset.py]: Also, never use `/` for ints. Use `//` instead. And then you also don't need the `int` cast.

File: GeneratingDataset.py [hunk 3]
```diff
@@ -1051,16 +1078,65 @@ def get_audio_features(self, audio, sample_rate):
...
     return feature_data
 
   def get_feature_dimension(self):
     """
     :rtype: int
     """
-    return (self.with_delta + 1) * self.num_feature_filters
+    if self.join_frames is not None:
+      return ((self.with_delta + 1) * self.num_feature_filters) * self.join_frames
+    return ((self.with_delta + 1) * self.num_feature_filters)
```
Comment (albertz) [inline GeneratingDataset.py]: I think it's cleaner to not duplicate code, but to write this instead:
`return return ((self.with_delta + 1) * self.num_feature_filters) * (self.join_frames or 1)`

Comment (albertz) [reply GeneratingDataset.py]: (Also, remove the unnecessary brackets.)

File: GeneratingDataset.py (hunk 2 above)
Comment (JackTemaki) [reply GeneratingDataset.py]: When being a parameter no spaces are PEP8 conform so there was no error, but I can change it.

Comment (albertz) [reply GeneratingDataset.py]: I think the code should be consistent. Also, even if not in PEP8, I think this is standard Python code convention.

**Pull request merged**